[WIP] [Workflow] Choose which Workflow events should be dispatched

stewartmalik opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #36617 
| License       | MIT
| Doc PR        | WIP

Allow users to specify which events should be dispatched when calling `workflow->apply()`.

---

- [x] still requires configuration method to be discussed
- [ ] submit changes to the documentation

stewartmalik committed:
[Workflow] Choose which Workflow events should be dispatched

stewartmalik committed:
[Workflow] Removed previous change regarding not firing the `announce` event.

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -309,7 +305,7 @@ private function leave(object $subject, Transition $transition, Marking $marking
         $places = $transition->getFroms();
 
-        if (null !== $this->dispatcher) {
+        if (null !== $this->dispatcher && $this->shouldDispatchEvent(WorkflowEvents::LEAVE)) {
```
what about include the `null !== $this->dispatcher` inside the method?

noniagriconomie commented on src/Symfony/Component/Workflow/Definition.php:
```diff
@@ -26,13 +26,15 @@ final class Definition
     private $transitions = [];
     private $initialPlaces = [];
     private $metadataStore;
+    private $dispatchEvents;
```
```suggestion
    private $dispatchedEvents;
```

??

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -407,4 +403,27 @@ private function announce(object $subject, Transition $initialTransition, Markin
+        }
+
+        // Check if the WorkflowEvent name is in the events that
+        // should be dispatched for this Workflow
```
```suggestion
        // should be dispatched for this transition
```

?

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -407,4 +403,27 @@ private function announce(object $subject, Transition $initialTransition, Markin
+        // Check if the WorkflowEvent name is in the events that
+        // should be dispatched for this Workflow
+        if (count($dispatchEvents) >= 1) {
+            return in_array($eventName, $dispatchEvents);
```
what about defining the third param if in_array just to be safe?

noniagriconomie commented on src/Symfony/Component/Workflow/CHANGELOG.md:
```diff
@@ -6,7 +6,7 @@ CHANGELOG
  * Added context to `TransitionException` and its child classes whenever they are thrown in `Workflow::apply()`
  * Added `Registry::has()` to check if a workflow exists
- * Added support for `$context[Workflow::DISABLE_ANNOUNCE_EVENT] = true` when calling `workflow->apply()` to not fire the announce event
+ * Added support for specifying which events should be dispatched when calling `workflow->apply()`
```
💎

lyrixx replied:
Indeed, this is a BC Break.

IHMO, you should merge the default configuration (the one passed to the construct) with the one in the `$context`

noniagriconomie commented on src/Symfony/Component/Workflow/WorkflowEvents.php:
```diff
@@ -55,4 +55,9 @@ final class WorkflowEvents
     {
     }
+
+    public static function getDefaultDispatchEvents(): array
```
```suggestion
    public static function getDefaultDispatchedEvents(): array
```

if you agree to rewrite the property above

stewartmalik committed:
[Workflow] Add feedback from Code Review. Add configuration support.

stewartmalik commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -324,6 +326,22 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
                                         ->defaultValue([])
                                         ->prototype('scalar')->end()
                                     ->end()
+                                    ->arrayNode('dispatched_events')
```
Could definitely use input on configuration here, this is my first time working on the Symfony code base and this was somewhat confusing.

I ran into issues with a number of things such as expressing an empty array in the XML configuration format, as well as having a default null value for arrayNodes.

I have worked around this by specifying `all` and `none` as allowed values for this configuration option however these are mostly hidden from the end user. The only exception is with the XML configuration where you need to specify `none` to remove all events from being fired.

stewartmalik committed:
Update `Workflow` to `transition`

Events are fired for Transitions however configured Workflow wide. Transition does make more sense in this instance.

Co-authored-by: Antoine Makdessi <antoine.makdessi@agriconomie.com>

stewartmalik committed:
[Workflow] Style fixes suggested by CI

stewartmalik committed:
[Workflow] Fix merge conflict

stewartmalik committed:
[Workflow] Another code style fix after the first one broke on patch download

stewartmalik commented:
Will be rebased when ready for final review.

fabpot commented:
/cc @lyrixx

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -363,48 +359,70 @@ private function enter(object $subject, Transition $transition, Marking $marking
-    private function announce(object $subject, Transition $initialTransition, Marking $marking): void
+    private function shouldDispatchEvent(string $eventName): bool
     {
+        // If we don't have a dispatcher we can't dispatch the
```
I think this comment is a little useless
Also if you use constants for next comparaison, comments may also be useless

lyrixx replied:
I agree

noniagriconomie reviewed (commented):
Some minor comments :)

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -324,6 +325,22 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
                                         ->prototype('scalar')->end()
                                     ->end()
+                                    ->arrayNode('dispatched_events')
+                                        ->beforeNormalization()
```
I don't think we should handle many way to configure the same thing.
to me, the perfect config would be:

1. no configuration => all
1. an empty array => non
1. an array with few items => only theses items
---
So I would remove the case for string. And it would allow to remove special cases for `all` and `none`.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -713,13 +713,25 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
 
+            // Record which events should be dispatched
+            if ($workflow['dispatched_events'] === ['all']) {
+                $dispatchedEvents = null;
```
See my previous comment

lyrixx commented on src/Symfony/Component/Workflow/Definition.php:
```diff
@@ -22,6 +22,16 @@
+     * Passing an array with WorkflowEvents will allow only those events to be dispatched plus
+     * the Guard Event.
+     *
+     * @var array|string[]
```
inded :+1:

and please, move this property after $metadataStore (I always sort the property declaration in the same order as the construct arguments)

lyrixx commented on src/Symfony/Component/Workflow/Definition.php:
```diff
@@ -32,7 +42,7 @@ final class Definition
      * @param string|string[]|null $initialPlaces
      */
-    public function __construct(array $places, array $transitions, $initialPlaces = null, MetadataStoreInterface $metadataStore = null)
+    public function __construct(array $places, array $transitions, $initialPlaces = null, MetadataStoreInterface $metadataStore = null, array $dispatchedEvents = null)
```
I don't think this should be hold by the definition. It's only related to the Workflow implementation, isn't ?

lyrixx reviewed (commented):
Sorry for the delay.

I left some comments about the design. Could you look at it?

Thanks

noniagriconomie commented:
@stewartmalik hello, do you need help to finish this one?
i would really use this feature :)
thx!

lyrixx commented:
For the record, I'm working on this PR (I'll reuse all the current, then cleaning it). I'll reopen a new PR ASAP

lyrixx commented:
Closing in favor of #37815 

@stewartmalik Thanks a lot for your first contribution, and for your issue. I opened a new PR, where I rebased and squashed your work. So don't worry for the ownership of the feature :)
